[Test Load Balancing] Validate `load-balance` option and enable to load test balanced

choheekim opened the pull request:
This pr contains two changes:

1) The weight-test-modules allows ordering a list of AMD modules by slowest to fastest test type where we assume that acceptance tests are the slowest test type, with integration, unit tests and lint test being fast and faster and fastest. Sorting the list of AMD modules will be functioned by default when load-balance option is specified to provide better utilization of browser parallelization.

2) With load-balance options enabled that the testLoader requires and unsees an AMD module as requested instead of testLoader loads a list of AMD modules all once. As the order of AMD modules ran per browser is not consistent with the load-balance option modules ran per browser are tracked and when all tests are complete it creates a json file which maps from a browser id to a list of AMD modules ran on the browser.

step2yeung commented on lib/commands/exam.js:
```diff
@@ -124,35 +135,90 @@ module.exports = TestCommand.extend({
-        }, []);
-      } else {
-        config.testPage = this._generateTestPages(baseUrl, partitions, count);
+    config.additionalBrowserSocketEvents = this._addBrowserSocketEventsToConfig(config);
```
shouldn't this only be done when `shouldLoadBalance` is true?

choheekim replied:
that's correct. Moved ` config.additionalBrowserSocketEvents = this._addBrowserSocketEventsToConfig(config);` under if (shouldLoadBalance) statement.

step2yeung commented on node-tests/acceptance/exam-test.js:
```diff
@@ -11,7 +11,15 @@ function getNumberOfTests(str) {
-var TOTAL_NUM_TESTS = 50;
+var TOTAL_NUM_TESTS = 53;
+
+function getTotalNumberOfTests(output) {
```
What if we pass `{ setupEmberOnerrorValidation: false }` to `start()` in ember-exam/tests/test-helper.js ?

step2yeung replied:
This got marked as resolved, but I don't see a change.
But to reiterate:
What if we pass `{ setupEmberOnerrorValidation: false }` to `start()` in `ember-exam/tests/test-helper.js`, then we would not need this method to modify TOTAL_NUM_TESTS depending on the number of browsers.

choheekim replied:
even if we pass { setupEmberOnerrorValidation: false } then getTotalNumberOfTests() will return 
 whatever number that's defined as TOTAL_NUM_TESTS.  getTotalNumberOfTests() parses test execution output file and subtract number of Ember.onError validation tests.

step2yeung commented on node-tests/unit/commands/exam-test.js:
```diff
@@ -198,6 +199,86 @@ describe('ExamCommand', function() {
+      ]);
+    });
+
+    it('should have a custom test page', function() {
```
Is this something new?

choheekim replied:
yes it's testing against load-balance option if testLoader modifies testem config with load-balance

scalvert commented on lib/commands/exam.js:
```diff
@@ -35,11 +37,17 @@ module.exports = TestCommand.extend({
     { name: 'partition',    type: [Array, Number],         description: 'The number of the partition(s) to run after splitting.' },
-    { name: 'weighted',     type: Boolean,                 description: 'Weights the type of tests to help equal splits.' },
     { name: 'parallel',     type: Boolean, default: false, description: 'Runs your split tests on parallel child processes.' },
+    { name: 'load-balance', type: Number,                  description: 'Loads your test balanced' },
```
That description isn't very informative. How about `Uses weighted load balancing to load tests in weighted order, slowest (acceptance) to fastest (lint).`

scalvert commented on lib/commands/exam.js:
```diff
@@ -91,7 +99,10 @@ module.exports = TestCommand.extend({
-      commandOptions.query = addToQuery(commandOptions.query, '_weighted', commandOptions.weighted);
+    }
+
+    if(this.validator.shouldLoadBalance) {
```
Nit: space after `if`

scalvert commented on lib/commands/exam.js:
```diff
@@ -91,7 +99,10 @@ module.exports = TestCommand.extend({
+    }
+
+    if(this.validator.shouldLoadBalance) {
+      commandOptions.query = addToQuery(commandOptions.query, '_load-balance', commandOptions.loadBalance);
```
Is it standard to use underscores for the query params? eg. `_load-balance`

scalvert commented on lib/commands/exam.js:
```diff
@@ -124,35 +135,89 @@ module.exports = TestCommand.extend({
+      const browserCount = commandOptions.loadBalance;
+      appendingParam = '_browser';
+
+      ids = Array.from({length: browserCount}, (v,k) => k+1); // if browser=3 then id = [1,2, 3]
```
I get what you're trying to say with this comment, but perhaps you can make it clearer? If you provide a value for browser, id becomes a range from 1..<browser>.

scalvert commented on lib/commands/exam.js:
```diff
@@ -124,35 +135,89 @@ module.exports = TestCommand.extend({
+   * @return {Object}
+   * @example tests/index.html?hidepassed&_split=3 if parallel
+   *          tests/index.html?hidepassed&_split=3&_load-balance if load-balance
+   *          tests/index.html?hidepassed&_split=3&_load-balance&_partition=1&_partition=2 if load-balance and partitions are specified
```
Again, I think you should ensure that using underscores is standard. If not, you should opt to go with no underscores for standardization.

choheekim replied:
As looking at ember-cli example underscores are used for a case like `load-balance`.
https://github.com/ember-cli/ember-cli/blob/master/lib/commands/test.js

scalvert commented on lib/commands/exam.js:
```diff
@@ -124,35 +135,89 @@ module.exports = TestCommand.extend({
+        baseUrl = addToUrl(baseUrl, '_split', splitCount);
+      } else if (command.validator.shouldLoadBalance) {
+        baseUrl = addToUrl(baseUrl, '_load-balance', true)
+        if (command.validator.shouldSplit) {
```
Couldn't this just be `||`ed with line 194?

scalvert commented on lib/commands/exam.js:
```diff
@@ -124,35 +135,89 @@ module.exports = TestCommand.extend({
+    let appendingParam = '_partition';
+
+    if (shouldParallelize && !ids) {
+      ids = [];
```
This is actually `browserIds`, isn't it? Should we rename this for clarity?

scalvert commented on lib/commands/exam.js:
```diff
@@ -254,5 +316,70 @@ module.exports = TestCommand.extend({
+      const fileName = `test-execution-${Date.now()}.json`;
+      const numOfBrowsersOfModuleMap = Object.keys(proto.moduleMap).length;
+
+      if(proto.finishedBrowsers === undefined) {
```
Nit: space after `if`.

scalvert commented on lib/commands/task/test-server.js:
```diff
@@ -0,0 +1,10 @@
+    testemOptions.additionalBrowserSocketEvents = options.additionalBrowserSocketEvents;
+    return testemOptions;
+  }
+});
```
Needs newline.

scalvert commented on lib/commands/exam.js:
```diff
@@ -254,5 +316,70 @@ module.exports = TestCommand.extend({
+    return events;
   }
-});
+});
```
Needs newline.

scalvert commented on addon-test-support/-private/weight-test-modules.js:
```diff
@@ -0,0 +1,37 @@
+                  const sortedModuleArr = groups[weight].sort();
+                  return accumulatedArray.concat(sortedModuleArr);
+              }, []);
+}
```
Needs newline.

scalvert commented on lib/commands/task/test.js:
```diff
@@ -0,0 +1,10 @@
+    testemOptions.additionalBrowserSocketEvents = options.additionalBrowserSocketEvents;
+    return testemOptions;
+  }
+});
```
Needs newline.

scalvert commented on lib/utils/tests-options-validator.js:
```diff
@@ -137,4 +137,24 @@ Object.defineProperty(TestsOptionsValidator.prototype, 'shouldParallelize', {
+  get: function _shouldLoadBalance() {
+    var loadBalance = this.options.loadBalance;
+
+    if(typeof loadBalance == 'undefined') {
```
Nit: space after `if`.

scalvert commented on lib/utils/tests-options-validator.js:
```diff
@@ -137,4 +137,24 @@ Object.defineProperty(TestsOptionsValidator.prototype, 'shouldParallelize', {
+  }
+});
+
+module.exports = TestsOptionsValidator;
```
Needs newline.

scalvert commented on node-tests/unit/commands/exam-test.js:
```diff
@@ -61,6 +61,12 @@ describe('ExamCommand', function() {
       });
     });
 
+    it('should set \'load-balance\' on the query option', function() {
```
"in the query options"

scalvert commented on tests/unit/weight-test-modules-test.js:
```diff
@@ -0,0 +1,15 @@
+
+module('Unit | weight-test-modules', () => {
+  test('should sort a list of file paths by weight', function(assert) {
+    const listOfModules = ['/jshint/test-1-test', '/acceptance/test-1-test', '/unit/test-1-test', '/integration/test-1-test', 'test-1-test']
```
Can we format this so each item is on a newline? This makes it _much_ easier to scan top to bottom rather than left to right. Same goes for all below.

choheekim commented on addon-test-support/-private/weight-test-modules.js:
```diff
@@ -0,0 +1,37 @@
+}
+
+const WEIGHT_MAP = {jshint:1, unit:10, integration:20, acceptance:150};
+const WEIGHT_DEFAULT = 50;
```
As a list of AMD modules we are getting is module file paths default_weight is for a file path which doesn't contain any of type of tests list in `WEIGHT_MAP`.

choheekim commented on addon-test-support/-private/patch-test-loader.js:
```diff
@@ -35,12 +38,31 @@ export default function patchTestLoader(TestLoader) {
 
-    const splitModules = splitTestModules(testLoader._testModules, split, partitions);
+    if (loadBalance) {
+      this.modules = weightTestModules(this.modules);
```
it doesn't update the `this.modules` as it only iterates `this.modules` and create new object to assign to `this.modules`.

step2yeung commented on node-tests/unit/utils/tests-options-validator-test.js:
```diff
@@ -117,4 +117,40 @@ describe('TestOptionsValidator', function() {
+      shouldThrow('ReplayExecution', { replayExecution: 'test-execution-0000000.json', replayBrowser: [1, 2, 1]}, /You cannot specify the same replayBrowser value twice. 1 is repeated./);
+    })
+
+    it('should throw an error if `replay-browser` contains duplicate values', function() {
```
the name of this test does not match the thing its testing

step2yeung commented on lib/commands/exam.js:
```diff
@@ -254,5 +370,87 @@ module.exports = TestCommand.extend({
+
+      // keep track of the modules ran per browserId when appMode is ci and replayExecution flag is false
+      if (this.config.appMode === 'ci' && !replayExecution) {
+        const browserId = /_browser=\s*(.*)/.exec(this.launcher.settings.test_page)[1];
```
I would suggest using `/_browser=\s*([0-9]*)/`. This would only get the numbers after `_browser`
Also, can we throw an error if browserId is `undefined`.

step2yeung commented on lib/commands/exam.js:
```diff
@@ -254,5 +370,87 @@ module.exports = TestCommand.extend({
+      }
+
+      // keep track of the modules ran per browserId when appMode is ci and replayExecution flag is false
+      if (this.config.appMode === 'ci' && !replayExecution) {
```
I feel this if should be inside the `else` condition in line 410, because if `moduleName` is null or undefined, we shouldn't need to write it to the moduleMap

step2yeung commented on lib/commands/task/test-server.js:
```diff
@@ -0,0 +1,12 @@
+
+module.exports = TestServerTask.extend({
+  
+  testemOptions(options) {
```
since ember-cli 3.2.0-beta1, this changed to `transformOptions`

step2yeung commented on lib/commands/task/test.js:
```diff
@@ -0,0 +1,12 @@
+
+module.exports = TestTask.extend({
+
+  testemOptions(options) {
```
same here

step2yeung commented on addon-test-support/load.js:
```diff
@@ -19,4 +22,60 @@ export default function loadEmberExam() {
+  }
+  qunit.begin(function() {
+    return new Promise(function(resolve) {
+      Testem.once('next-module-response', function getTestModule(moduleName) {
```
incase the `--filter` option is passed, this won't load tests from the module if none of the test name matches the filter. If this happens, it needs to continue emit `get-next-module`

This should look something like:
```
Testem.on('next-module-response', function getTestModule(moduleName) {
        if (moduleName !== undefined) {
          loadIndividualModule(moduleName);
        }

        // if no tests were added, request the next module
        if (QUnit.config.queue.length === 0) {
          Testem.emit('get-next-module');
        } else {
          Testem.removeDuplicateCallbacks('next-module-response', getTestModule);
          resolve();
        }
      });
```

step2yeung commented on addon-test-support/load.js:
```diff
@@ -19,4 +22,60 @@ export default function loadEmberExam() {
+      Testem.emit('get-next-module');
+    });
+  });
+  qunit.moduleDone(function() {
```
similar thing here.
one difference, we can probably do without `resolveCallbackFn` and use resolve directly.
```
qunit.moduleDone(function() {
    return new Promise(function(resolve) {
      Testem.on('next-module-response', function getTestModule(moduleName) {
        if (moduleName !== undefined) {
          loadIndividualModule(moduleName);
        }

        // if no tests were added, request the next module
        if (QUnit.config.queue.length === 0) {
          Testem.emit('get-next-module');
        } else {
          Testem.removeDuplicateCallbacks('next-module-response', getTestModule);
          Testem.removeDuplicateCallbacks('module-queue-complete', resolve);
          resolve();
        }
      });
      Testem.on('module-queue-complete', resolve);
      Testem.emit('get-next-module');
    });
  });
```

stefanpenner commented on addon-test-support/load.js:
```diff
@@ -19,4 +22,69 @@ export default function loadEmberExam() {
+          Testem.emit('get-next-module');
+        } else {
+          Testem.removeEventCallbacks('next-module-response', getTestModule);
+          resolve();
```
This resolve may not always be called, if anything in this callback throws, we simply wont invoke this resolve nor do we reject.

choheekim replied:
`loadIndividualModule` is the callback resolves the promises. The function calls `testLoader.loadIndividualModule` that’s in `patch-test-loader` and  `testLoader.loadIndividualModule` requires and unsess a given module. It seems like `require` and `unsee` already handles an error that might be thrown. I’m thinking `loadIndividualModule` won’t cause execution stops or unresolves the promise.
https://github.com/ember-cli/ember-cli-test-loader/blob/master/addon-test-support/index.js#L71

scalvert replied:
Would it at least be prudent to `reject` if errors occur?

stefanpenner commented on addon-test-support/load.js:
```diff
@@ -19,4 +22,69 @@ export default function loadEmberExam() {
+  }
+  qunit.begin(function() {
+    return new Promise(function(resolve) {
+      Testem.once('next-module-response', function getTestModule(moduleName) {
```
what happens if this event fails? For example, what happens if loading a test module causes an error to be thrown, how does the test runner proceed?

stefanpenner commented on addon-test-support/-private/patch-test-loader.js:
```diff
@@ -33,14 +36,32 @@ export default function patchTestLoader(TestLoader) {
+    this.modules = splitTestModules(this.modules, split, partitions);
+
+    if (loadBalance) {
+      Testem.emit('set-modules-queue', this.modules);
```
Why emit an event, rather then invoke a method?

scalvert replied:
So essentially this is performing the communication between the browser and loader, and needs to send the modules through the socket to the browser to know what modules are to be queued, correct?

choheekim replied:
It's sending the modules to server through the socket. server becomes a source of truth for the modules and browsers ask the server by emitting `get-next-module` event to get next module to run.

https://github.com/ember-cli/ember-exam/pull/136/files/c30c1e99e635910e7e21a01546534ae1a010a8e0#diff-babb5d4fe8ce7520c798eaae9f5393a0R54

stefanpenner commented on lib/commands/exam.js:
```diff
@@ -124,35 +180,98 @@ module.exports = TestCommand.extend({
+   *          tests/index.html?hidepassed&_split=3&_loadBalance if load-balance
+   *          tests/index.html?hidepassed&_split=3&_loadBalance&_partition=1&_partition=2 if load-balance and partitions are specified
+   */
+  _getCustomBaseUrl: function(config, commandOptions) {
```
this seems a tad complicated, can we unit test?

choheekim replied:
`getCustomBaseUrl` is used in `generateCustomConfigs` and we do have unit tests for `generateCustomConfigs`. do you still think that unit tests for  `getCustomBaseUrl` would be helpful? fyi the unit tests are in `node-tests/unit/commands/exam-test.js`
https://github.com/ember-cli/ember-exam/pull/136/files#r229002197

scalvert replied:
Having a bit of redundancy in tests is OK imo, as it helps prevent regressions that happen as a result of other changes in the system. I'd add tests.

choheekim replied:
Added tests but it will fail until #148 landed. It verifies to have ember-cli@3.2 above for load-balance option. Since in this pr ember-cli is lower version tests that does option validator for load-balance will be failed.

stefanpenner commented on lib/commands/exam.js:
```diff
@@ -254,5 +370,91 @@ module.exports = TestCommand.extend({
+        delete proto.finishedBrowser;
+        if (this.config.appMode === 'ci') {
+          const testExecutionJson = JSON.stringify(moduleMapJsonObj);
+          fs.writeFile(fileName, testExecutionJson, (err) => {
```
why async IO, rather then sync? async IO although non blocking, does increase complexity re: concurrency.

step2yeung commented on lib/commands/exam.js:
```diff
@@ -254,5 +370,91 @@ module.exports = TestCommand.extend({
+
+      // Updating browserModuleMapping which maps from browser ids to a list of AMD modules ran per the browser
+      // is needed when running test suite in ci.
+      if (this.config.appMode === 'ci' && browserModuleMapping && !this.module) {
```
i don't see `this.module` defined... do you mean `this.moduleQueue`?

step2yeung commented on lib/commands/exam.js:
```diff
@@ -124,35 +180,98 @@ module.exports = TestCommand.extend({
+    } else if (shouldReplayExecution) {
+      const executionMapping = this.getExecutionMappigInstance(commandOptions);
+      config.browser_module_mapping = executionMapping['executionMapping'];
+
```
We will also need 
`config.additional_browser_socket_events = this._addLoadBalancingBrowserSocketEvents(config);`
in this if condition, since it still requires the socket events to be setup

step2yeung commented on lib/commands/exam.js:
```diff
@@ -124,35 +180,98 @@ module.exports = TestCommand.extend({
+      const executionMapping = this.getExecutionMappigInstance(commandOptions);
+      config.browser_module_mapping = executionMapping['executionMapping'];
+
+      appendingParam = '_browser';
```
in the case for `shouldReplayExecution`
the testPage will look something like: `[ 'tests/index.html?hidepassed=true&_browser=1' ]`
without the `_loadBalance` param, the browser won't fire the `setModuleQueue` event on patch-test-loader.js line 48

choheekim replied:
As we are running `_loadBalance` as default for `replay-execution` what about adding `_loadBalance` to the url so that the event can get fired.

step2yeung commented on lib/commands/exam.js:
```diff
@@ -254,5 +370,91 @@ module.exports = TestCommand.extend({
+        // TODO: Investigate not using proto
+        delete proto.moduleQueue;
+        delete proto.finishedBrowser;
+        if (this.config.appMode === 'ci') {
```
we should also check for`!replayExecution` as well, since we won't want to write the execution file when we are rerunning tests

choheekim replied:
I guess we don't even need to register `after-tests-complete` event when it's running for `replay-execution` because `after-tests-complete` is just to create a json file. I'm checking if it's `replay-execution` and if it is then do not register the event at all.

stefanpenner commented:
I will attempt to make time tomorrow to do a thorough review.

scalvert commented on addon-test-support/load.js:
```diff
@@ -19,4 +22,69 @@ export default function loadEmberExam() {
+
+// loadIndividualModule() executes loadIndividualModule() defined in patch-test-loader. It enables to load an AMD module one by one.
+function loadIndividualModule(moduleName) {
+  return testLoader.loadIndividualModule(moduleName);
```
Does this function need to return at all? The return value is never used AFAICT.

scalvert commented on lib/commands/exam.js:
```diff
@@ -101,6 +116,47 @@ module.exports = TestCommand.extend({
+   * @param {Object} option
+   */
+  getExecutionMappigInstance: function(options) {
+    if (executionMappingObj !== undefined && executionMappingObj['numberOfBrowsers'] !== undefined && executionMappingObj['executionMapping'] !== undefined) {
```
Could we not use Hungarian notation here for `executionMappingObj`? I think just dropping the `Obj` would suffice.

scalvert commented on lib/commands/exam.js:
```diff
@@ -3,8 +3,12 @@
+var testTask = require('./task/test');
 var debug = require('debug')('exam');
 
+const executionMappingObj = {};
```
Is this intended to be used as a dictionary? If so, you may want to consider using `Object.create(null)` so it doesn't inherit.

scalvert commented on lib/commands/exam.js:
```diff
@@ -101,6 +116,47 @@ module.exports = TestCommand.extend({
+    const filePath = options.replayExecution;
+    const replayBrowser = options.replayBrowser;
+
+    let testMappingObj = {};
```
Again, Hungarian is probably not necessary here.

scalvert commented on lib/commands/exam.js:
```diff
@@ -101,6 +116,47 @@ module.exports = TestCommand.extend({
+    if (executionMappingObj !== undefined && executionMappingObj['numberOfBrowsers'] !== undefined && executionMappingObj['executionMapping'] !== undefined) {
+      return executionMappingObj;
+    }
+    const filePath = options.replayExecution;
```
`const executionFilePath =` is a bit more descriptive here.

scalvert commented on lib/commands/exam.js:
```diff
@@ -101,6 +116,47 @@ module.exports = TestCommand.extend({
+      return executionMappingObj;
+    }
+    const filePath = options.replayExecution;
+    const replayBrowser = options.replayBrowser;
```
`const browserIdsToReplay` may be a bit more meaningful.

scalvert commented on lib/commands/exam.js:
```diff
@@ -101,6 +116,47 @@ module.exports = TestCommand.extend({
+
+    try {
+      // Read the replay execution json file.
+      const testMappingRawData = fs.readFileSync(filePath);
```
`const executionToReplay =` seems more accurate, no?

scalvert commented on lib/commands/exam.js:
```diff
@@ -101,6 +116,47 @@ module.exports = TestCommand.extend({
+      const testMappingRawData = fs.readFileSync(filePath);
+      testMappingObj = JSON.parse(testMappingRawData);
+    } catch (err) {
+      const errorMsg = 'Error reading reply execution JSON file - ' + err;
```
You can probably just inline this into the `Error` constructor.

scalvert commented on lib/commands/exam.js:
```diff
@@ -101,6 +116,47 @@ module.exports = TestCommand.extend({
+    executionMappingObj['executionMapping'] = ((replayBrowser, browserModuleMappingObj) => {
+      const browserModuleMap = {};
+
+      for (const idx in replayBrowser) {
```
`for (const browserId in browserIdsToReplay) {`, if you make the suggested renaming above.

scalvert commented on lib/commands/exam.js:
```diff
@@ -101,6 +116,47 @@ module.exports = TestCommand.extend({
+      const browserModuleMap = {};
+
+      for (const idx in replayBrowser) {
+        const browserId = replayBrowser[idx];
```
I'm a bit confused by this, as `replayBrowser` is defined as this in the options:

> { name: 'replay-browser',   type: [Array, Number],          description: 'The number of the browser(s) to run from a specified file path'}

Is this either an `Array<int>` _or_ an `Array<Array<int>>`?

choheekim replied:
`replayBrowser` is an `Array<int>`. in `for .. in` a value of idx is an index of replayBrowser. for instance, replayBrowser = [1, 2, 3] then in the first iteration of the for loop idx value will be 0 and browserId will be replayBrowser[0]. To reduce confusion here what about using `for( ; ; ;)` to iterate an array?

scalvert commented on lib/commands/exam.js:
```diff
@@ -124,35 +180,100 @@ module.exports = TestCommand.extend({
+      const browserCount = commandOptions.loadBalance;
+      appendingParam = 'browser';
+
+      // Creates an array starting from 1 to browserCount. e.g. if brownserCount is 3, the returned array is [1, 2, 3]
```
> e.g. if brownserCount is 3

Typo: e.g. if *browserCount* is 3

scalvert commented on addon-test-support/-private/weight-test-modules.js:
```diff
@@ -0,0 +1,37 @@
+const WEIGHT_MAP = {jshint:1, eslint:1, unit:10, integration:20, acceptance:150};
```
Can we change this object into a more readable format? Also, do we need `jshint`?

scalvert replied:
Also, how about we call this `TEST_TYPE_WEIGHT`? It's not really relevant that it's a map, is it?

choheekim replied:
why do you think we don't need `jshint`?

scalvert replied:
Why do you think we do? Is this for backwards compatibility for older blueprints?

scalvert commented on addon-test-support/-private/weight-test-modules.js:
```diff
@@ -0,0 +1,37 @@
+const WEIGHT_MAP = {jshint:1, eslint:1, unit:10, integration:20, acceptance:150};
+const WEIGHT_DEFAULT = 50;
```
Can we call this `DEFAULT_WEIGHT`? It reads more like english that way. I think I mentioned this before in a previous comment.

scalvert commented on addon-test-support/-private/weight-test-modules.js:
```diff
@@ -0,0 +1,37 @@
+* @param {*} modulePath File path to a module
+ */
+function getWeight(modulePath) {
+  const [_,key] = /\/(jshint|unit|integration|acceptance)\//.exec(modulePath) || [];
```
Can we reformat this destructure so it's more readable?

```js
const [_, key] = ...
```

scalvert commented on addon-test-support/load.js:
```diff
@@ -19,4 +22,69 @@ export default function loadEmberExam() {
+  }
+  qunit.begin(function() {
+    return new Promise(function(resolve) {
+      Testem.on('testem:next-module-response', function getTestModule(moduleName) {
```
Not a huge thing, and you can totally ignore this, but what about using something like [`p-event`](https://www.npmjs.com/package/p-event) here?

scalvert commented on lib/commands/exam.js:
```diff
@@ -101,6 +116,47 @@ module.exports = TestCommand.extend({
 
+  /**
+   *
+   * Instantiate an executionMappingObj object
```
Let's remove this comment. It's not useful.

scalvert commented on lib/commands/exam.js:
```diff
@@ -101,6 +116,47 @@ module.exports = TestCommand.extend({
+    executionMapping['executionMapping'] = ((browserIdsToReplay, browserModuleMapping) => {
+      const browserModuleMap = {};
+
+      for(var i = 0; i< browserIdsToReplay.length; i++) {
```
Can we format this so the the spacing is cleaner?

```js
for (var i = 0; i < browserIdsToReplay.length; i++) {
```

scalvert commented on lib/commands/exam.js:
```diff
@@ -254,5 +372,97 @@ module.exports = TestCommand.extend({
+        delete proto.moduleQueue;
+        delete proto.finishedBrowser;
+        if (this.config.appMode === 'ci' && shouldLoadBalance) {
+          const fileName = `test-execution-${Date.now()}.json`;
```
I realize that this may be unique, but it doesn't help anyone identify an execution easily. Can we instead format this into a more human readable format, and one that will be sortable by time of execution so it's easier to identify a serialized execution that you may want to rerun?

choheekim commented on lib/utils/tests-options-validator.js:
```diff
@@ -52,7 +51,35 @@ Object.defineProperty(TestsOptionsValidator.prototype, 'shouldSplit', {
+ */
+function validateReplayBrowser(replayBrowser, replayExecution, options) {
+
+  exam.prototype.executionMappingObj = exam.prototype.getExecutionMappigInstance(options);
```
We don't have tests for `validateReplayBrowser` but we have tests for validator which tests any violations of option usage that covers cases in the function.

step2yeung commented on lib/commands/exam.js:
```diff
@@ -101,6 +117,45 @@ module.exports = TestCommand.extend({
+    const browserModuleMapping = testMapping.executionMapping;
+
+    executionMapping.numberOfBrowsers = testMapping.numberOfBrowsers;
+    executionMapping.executionMapping = ((browserIdsToReplay, browserModuleMapping) => {
```
`executionMapping.executionMapping` seems a bit weird to me, since the name of the attribute is the same as the owner/parent.
How about `executionMapping.browserToModuleMap`?

stefanpenner commented:
I am partway through rebasing this PR.

stefanpenner commented:
Just an update, I have:

* squash @choheekim's commits into 1 commit (making it easy to rebase)
* fixed tests
* rebased so this could be applied cleanly to master

stefanpenner commented on .travis.yml:
```diff
@@ -2,11 +2,20 @@
 language: node_js
 node_js:
   - "6"
+<<<<<<< HEAD
```
oops

stefanpenner commented on addon-test-support/load.js:
```diff
@@ -19,4 +22,73 @@ export default function loadEmberExam() {
+  }
+  qunit.begin(function() {
+    return new self.Promise(function(resolve) {
+      Testem.on('testem:next-module-response', function getTestModule(moduleName) {
```
the `begin` and `moduleDone` callbacks appear extremely similar, should they be unified?

stefanpenner commented on addon-test-support/load.js:
```diff
@@ -19,4 +22,73 @@ export default function loadEmberExam() {
+          resolve();
+        }
+      });
+      Testem.on('testem:module-queue-complete', resolve);
```
I believe we leak event listeners here, specifically: if `testem:module-queue-complete` is triggered without `next-module-response` being triggered, who will remove the `testem:next-module-response` when we are done with it?

stefanpenner commented on addon-test-support/load.js:
```diff
@@ -19,4 +22,73 @@ export default function loadEmberExam() {
+
+        // if no tests were added, request the next module
+        if (qunit.config.queue.length === 0) {
+          Testem.emit('get-next-module');
```
should this event also be prefixed with `testem`?

stefanpenner commented on lib/commands/exam.js:
```diff
@@ -125,34 +180,99 @@ module.exports = TestCommand.extend({
+
+    const command = this;
+
+    const appendParamToBaseUrl = function(baseUrl) {
```
can this be a standalone unit tested function?

stefanpenner commented on lib/commands/exam.js:
```diff
@@ -254,5 +372,97 @@ module.exports = TestCommand.extend({
+
+    events['set-modules-queue'] = function(modules) {
+      // When `load-balance` option is valid we want to have one static list of modules on server side to send a module path to browsers.
+      const proto = Object.getPrototypeOf(this);
```
Using the prototype for storing/sharing state between these two events is not appropriate. Prototypes are general reserved for inherited properties and methods.

If we have explicit state, lets create a new object for said state, with clear methods for interaction. That way it is explicit, obvious, maintainable and testable.

I'll wait until this area is refactored as described above before continuing the review.

step2yeung replied:
@stefanpenner the event handlers here gets registered on testem's browser runner's socket connection: https://github.com/testem/testem/commit/209920ff3e523dc74b1f22e98ac3415b59a2eae7
So the state needs to be saved within testem's scope.

Within testem, there is no official place to store test state. With `config` being a shared object that all runner instance has access to. Would using `config` for this purpose be reasonable?

step2yeung replied:
Resolved this by adding a execution-state-manager

stefanpenner commented on lib/commands/exam.js:
```diff
@@ -254,5 +372,97 @@ module.exports = TestCommand.extend({
+    }
+
+    events['get-next-module'] = function() {
+      const proto = Object.getPrototypeOf(this);
```
^

stefanpenner commented on lib/commands/exam.js:
```diff
@@ -254,5 +372,97 @@ module.exports = TestCommand.extend({
+  _addLoadBalancingBrowserSocketEvents(config) {
+    const events = config.custom_browser_socket_events || {};
+
+    events['set-modules-queue'] = function(modules) {
```
Should we prefix this event?

choheekim replied:
We wanted to distinguish events which are sent from server side to browser side and from browser side to server side. Events emitted in server side are prefixed with `testem`. If we prefix all events I think it's beneficial as it's all unified but wouldn't it give confusion where we have to distinguish where the events are sent from?

scalvert replied:
Why not prefix on both sides, rather than one side?

stefanpenner commented on lib/commands/exam.js:
```diff
@@ -254,5 +372,97 @@ module.exports = TestCommand.extend({
+      }
+    }
+
+    events['get-next-module'] = function() {
```
Should we prefix this event?

stefanpenner commented on lib/commands/exam.js:
```diff
@@ -254,5 +372,97 @@ module.exports = TestCommand.extend({
+            executionMapping: proto.moduleMap
+          };
+          const testExecutionJson = JSON.stringify(moduleMapJson);
+          fs.writeFileSync(fileName, testExecutionJson, (err) => {
```
`fs.writeFileSync` does not have a 3rd callback for error handling: https://nodejs.org/api/fs.html#fs_fs_writefilesync_file_data_options

stefanpenner commented on lib/commands/exam.js:
```diff
@@ -254,5 +372,97 @@ module.exports = TestCommand.extend({
+        delete proto.moduleQueue;
+        delete proto.finishedBrowser;
+        if (this.config.appMode === 'ci' && shouldLoadBalance) {
+          const fileName = `test-execution-${moment().format('YYYY-MM-DD__HH-MM-SS')}.json`;
```
moment is a pretty heavy dependency for this, could I recommend instead this value is simply the raw `Date.now()` or `new Date().toJSON().replace(/[:.]/g,'-')`

scalvert replied:
+1

stefanpenner commented:
So, although I have lots of feedback. Please don't take it the wrong way, I believe we should continue to work on this effort, as it will be worthwhile. My only goal, is to ensure it is as maintainable and simple as possible :)

stefanpenner commented on tests/unit/qunit/test-loader-test.js:
```diff
@@ -19,20 +19,19 @@ module('Unit | test-loader', {
-
-    this.originalURLParams = TestLoader._urlParams;
+    this.testLoader = new EmberExamTestLoader();
+    this.originalURLParams = EmberExamTestLoader._urlParams;
```
In general, we want `state` to be stored on instances and not as static class members. `_urlParams` is best described of as state, and as such should be instance state of the `testLoader` and not static state of its class.

stefanpenner replied:
In general, such static state is reserved for such things as constants which remain constant across all usage of the class, or static functions.

stefanpenner replied:
As implemented, it becomes "shared global state", and thus requires the bellow cleanup in `afterEach`.
Truthfully, the way it is used, really lends itself to be a parameter to the EmberExamTestLoader class.

stefanpenner commented on tests/unit/qunit/test-loader-test.js:
```diff
@@ -19,20 +19,19 @@ module('Unit | test-loader', {
-  TestLoader._urlParams = {};
-
-  TestLoader.load();
+  this.testLoader._urlParams = {};
```
`_` prefixed variables are considered private, mutating them external in tests is very unexpected. It is breaking encapsulation.  If we must changes these arguments on every test, lets simply instantiate a new `EmberExamTestLoader` class for each test.

```js
const testLoader = new EmberExamTestLoader({} /* urlParams */);

// or maybe:
const testLoader = new EmberExamTestLoader({ urlParams });
```

and then no cleanup is required, as no global pollution is possible.

stefanpenner commented on tests/unit/qunit/test-loader-test.js:
```diff
@@ -47,12 +46,11 @@ test('loads all test modules by default', function(assert) {
-  TestLoader._urlParams = {
-    _partition: 3,
-    _split: 4,
+  this.testLoader._urlParams = {
```
^

stefanpenner commented on tests/unit/qunit/test-loader-test.js:
```diff
@@ -61,12 +59,12 @@ test('loads modules from a specified partition', function(assert) {
-  TestLoader._urlParams = {
-    _partition: [1, 3],
-    _split: 4,
+  this.testLoader._urlParams = {
```
^

stefanpenner commented on node-tests/unit/commands/exam-test.js:
```diff
@@ -97,29 +105,116 @@ describe('ExamCommand', function() {
+
+      command.validator = new TestOptionsValidator(options);
+
+      return command._getCustomBaseUrl(config, options)
```
Testing private API in tests suggests some abstraction violation. Either it should be standalone and tested, or maybe it should be simply considered public API.

choheekim commented:
Work in progress addressing the comments
1. Create a customized emberTestLoader (done)
2. Resolve mixing usage of `this` and `testLoader` (done)
3. Add prefix `testem:` for every event (done)
4. Refactor QUnit begin & moduleDone callback (done)
 - As both has similar functionality simplify the code so that same code doesn't repeat
 - Handle/Reject a case where module is not defined or any error occurred during execution of the callbacks
5. Add unit tests for `appendParamToBaseUrl` (done)
6. Instantiate a new EmberExamTestLoader class for each test (done)
7. Refactor `exam.js` so that it's more testable (WIP)
8. Address not to use `prototype` in test-options-validator.js (done)

step2yeung commented:
Chohee and I worked on the refactoring, although there are some TODO spread around which we will get to before trying to land this, it should be in a good state for another run through :)

step2yeung commented:
> do we really need to bring in `moment` as a dependency just for this? can we use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString instead?

Good call @Turbo87, the usage has been removed, but wasn't removed from yarn.lock.

step2yeung commented:
Made a handle of changes:
- rebased against master
- further refactoring
- added acceptance tests invoking the new commands
- made setup simpler with `loadEmberExam()` 
- options changes: `parallel` accepts a number, `load-balance` is a boolean, `replay-browser` is optional to `replay-execution`
- when browser crashes, the test-execution json will still be written to disk

step2yeung commented on lib/commands/exam.js:
```diff
@@ -120,141 +124,97 @@ module.exports = TestCommand.extend({
+        }
+      };
+
+      events['test-start'] = function() {
```
we should not used `test-start` event for this.

stefanpenner commented:
Running this locally, 

```sh
 ember exam --path=dist  --split=4 --partition=1 --load-balance --parallel=4
```

Unexpectedly, didn't seem to run tests in the expected order. Specifically, I expected Acceptance tests to run first? 

```sh
[3:28:34 PM] ok 1 Chrome 71.0 - [349 ms] - Exam Partition 1 - Integration 
[3:28:34 PM] ok 2 Chrome 71.0 - [181 ms] - Exam Partition 1 - Integration 
[3:28:35 PM] ok 3 Chrome 71.0 - [258 ms] - Exam Partition 1 - Integration 
[3:28:35 PM] ok 4 Chrome 71.0 - [243 ms] - Exam Partition 1 - Integration 
[3:28:35 PM] ok 5 Chrome 71.0 - [164 ms] - Exam Partition 1 - Integration 
[3:28:35 PM] ok 6 Chrome 71.0 - [155 ms] - Exam Partition 1 - Integration 
[3:28:35 PM] ok 7 Chrome 71.0 - [170 ms] - Exam Partition 1 - Integration 
[3:28:35 PM] ok 8 Chrome 71.0 - [186 ms] - Exam Partition 1 - Integration 
[3:28:36 PM] ok 9 Chrome 71.0 - [200 ms] - Exam Partition 1 - Integration 
[3:28:36 PM] ok 10 Chrome 71.0 - [188 ms] - Exam Partition 1 - Integration
[3:28:36 PM] ok 11 Chrome 71.0 - [195 ms] - Exam Partition 1 - Integration
[3:28:36 PM] ok 12 Chrome 71.0 - [182 ms] - Exam Partition 1 - Integration
[3:28:36 PM] ok 13 Chrome 71.0 - [164 ms] - Exam Partition 1 - Integration
[3:28:37 PM] ok 14 Chrome 71.0 - [166 ms] - Exam Partition 1 - Integration
[3:28:37 PM] ok 15 Chrome 71.0 - [186 ms] - Exam Partition 1 - Integration
[3:28:37 PM] ok 16 Chrome 71.0 - [170 ms] - Exam Partition 1 - Integration
[3:28:37 PM] ok 17 Chrome 71.0 - [222 ms] - Exam Partition 1 - Unit
[3:28:37 PM] ok 18 Chrome 71.0 - [176 ms] - Exam Partition 1 - Unit
[3:28:37 PM] ok 19 Chrome 71.0 - [209 ms] - Exam Partition 1 - Unit
[3:28:38 PM] ok 20 Chrome 71.0 - [668 ms] - Exam Partition 1 - Unit
[3:28:38 PM] ok 21 Chrome 71.0 - [181 ms] - Exam Partition 1 - Unit
[3:28:39 PM] ok 22 Chrome 71.0 - [203 ms] - Exam Partition 1 - Unit
[3:28:39 PM] ok 23 Chrome 71.0 - [198 ms] - Exam Partition 1 - Unit
[3:28:40 PM] ok 24 Chrome 71.0 - [666 ms] - Exam Partition 1 - Unit
3:28:43 PM] ok 25 Chrome 71.0 - [3333 ms] - Exam Partition 1 - Acceptance 
[3:28:44 PM] ok 26 Chrome 71.0 - [829 ms] - Exam Partition 1 - Acceptance 
[3:28:44 PM] ok 27 Chrome 71.0 - [639 ms] - Exam Partition 1 - Acceptance 
[3:28:45 PM] ok 28 Chrome 71.0 - [622 ms] - Exam Partition 1 - Acceptance 
[3:28:46 PM] ok 29 Chrome 71.0 - [621 ms] - Exam Partition 1 - Acceptance 
[3:28:46 PM] ok 30 Chrome 71.0 - [708 ms] - Exam Partition 1 - Acceptance 
[3:28:47 PM] ok 31 Chrome 71.0 - [724 ms] - Exam Partition 1 - Acceptance 
[3:28:48 PM] ok 32 Chrome 71.0 - [796 ms] - Exam Partition 1 - Acceptance 
[3:28:50 PM] ok 33 Chrome 71.0 - [2335 ms] - Exam Partition 1 - Acceptance
[3:28:53 PM] ok 34 Chrome 71.0 - [2471 ms] - Exam Partition 1 - Acceptance
[3:28:56 PM] ok 35 Chrome 71.0 - [2666 ms] - Exam Partition 1 - Acceptance
[3:28:59 PM] ok 36 Chrome 71.0 - [3684 ms] - Exam Partition 1 - Acceptance
[3:29:02 PM] ok 37 Chrome 71.0 - [2462 ms] - Exam Partition 1 - Acceptance
```
Am I misunderstanding how the weighted stuff should work?

stefanpenner commented:
oops my last comment may be invalid, I linked stuff wrong. Linking correctly yields:

```
3:38:13 PM] ok 1 Chrome 71.0 - [2285 ms] - <e23de2b1> - Acceptance 
[3:38:13 PM] ok 2 Chrome 71.0 - [200 ms] - <2429b676> - Integration
[3:38:13 PM] ok 3 Chrome 71.0 - [343 ms] - <f4e64863> - Integration
[3:38:13 PM] ok 4 Chrome 71.0 - [184 ms] - <97310d76> - Integration
[3:38:14 PM] ok 5 Chrome 71.0 - [179 ms] - <0fcefaa8> - Integration
[3:38:14 PM] ok 6 Chrome 71.0 - [173 ms] - <08933498> - Integration
[3:38:14 PM] ok 7 Chrome 71.0 - [185 ms] - <6182049f> - Integration
[3:38:14 PM] ok 8 Chrome 71.0 - [170 ms] - <ceac8b9c> - Integration
[3:38:14 PM] ok 9 Chrome 71.0 - [177 ms] - <5f985da4> - Integration
[3:38:15 PM] ok 10 Chrome 71.0 - [189 ms] - <b036fef5> - Integration
[3:38:15 PM] ok 11 Chrome 71.0 - [206 ms] - <edf9458e> - Integration
[3:38:15 PM] ok 12 Chrome 71.0 - [198 ms] - <9f70bb68> - Integration
[3:38:15 PM] ok 13 Chrome 71.0 - [210 ms] - <765d49e5> - Integration
[3:38:15 PM] ok 14 Chrome 71.0 - [209 ms] - <886390e2> - Integration
[3:38:16 PM] ok 15 Chrome 71.0 - [196 ms] - <5dae8899> - Integration
[3:38:16 PM] ok 16 Chrome 71.0 - [179 ms] - <34ff1d20> - Integration
[3:38:16 PM] ok 17 Chrome 71.0 - [188 ms] - <a7dab4a5> - Integration
[3:38:16 PM] ok 18 Chrome 71.0 - [194 ms] - <b5b02428> - Integration
[3:38:16 PM] ok 19 Chrome 71.0 - [167 ms] - <454b25e2> - Integration
[3:38:16 PM] ok 20 Chrome 71.0 - [167 ms] - <4196f0e5> - Integration
[3:38:17 PM] ok 21 Chrome 71.0 - [184 ms] - <bf05c0b3> - Integration
[3:38:17 PM] ok 22 Chrome 71.0 - [191 ms] - <184d4e06> - Integration
[3:38:17 PM] ok 23 Chrome 71.0 - [193 ms] - <c9b6f60e> - Integration
[3:38:17 PM] ok 24 Chrome 71.0 - [189 ms] - <2f773e15> - Integration
[3:38:17 PM] ok 25 Chrome 71.0 - [182 ms] - <e4ce6c5a> - Integration
[3:38:18 PM] ok 26 Chrome 71.0 - [187 ms] - <f9a4dc13> - Integration
[3:38:18 PM] ok 27 Chrome 71.0 - [169 ms] - <9efa5897> - Integration
[3:38:18 PM] ok 28 Chrome 71.0 - [190 ms] - <ae569e9a> - Integration
[3:38:18 PM] ok 29 Chrome 71.0 - [176 ms] - <69b31a28> - Integration
[3:38:18 PM] ok 30 Chrome 71.0 - [175 ms] - <c2faa77b> - Integration
[3:38:19 PM] ok 31 Chrome 71.0 - [193 ms] - <f38d4eed> - Unit
[3:38:19 PM] ok 32 Chrome 71.0 - [208 ms] - <9f4a1bd6> - Unit
[3:38:19 PM] ok 33 Chrome 71.0 - [288 ms] - <e192eaf4> - Unit
[3:38:19 PM] ok 34 Chrome 71.0 - [179 ms] - <d91e0285> - Unit
[3:38:19 PM] ok 35 Chrome 71.0 - [167 ms] - <910a780f> - Unit
[3:38:20 PM] ok 36 Chrome 71.0 - [160 ms] - <2e9fc642> - Unit
[3:38:20 PM] ok 37 Chrome 71.0 - [177 ms] - <1da75ea8> - Unit
[3:38:20 PM] ok 38 Chrome 71.0 - [173 ms] - <aefe5ddf> - Unit
[3:38:20 PM] ok 39 Chrome 71.0 - [163 ms] - <dd3be0ee> - Unit
[3:38:20 PM] ok 40 Chrome 71.0 - [165 ms] - <0316c9ba> - Unit
[3:38:20 PM] ok 41 Chrome 71.0 - [157 ms] - <f06f73f9> - Unit
[3:38:21 PM] ok 42 Chrome 71.0 - [167 ms] - <f1f1ffbb> - Unit
[3:38:21 PM] ok 43 Chrome 71.0 - [199 ms] - <fc2f94fd> - Unit
[3:38:21 PM] ok 44 Chrome 71.0 - [164 ms] - <27e95eff> - Unit
[3:38:21 PM] ok 45 Chrome 71.0 - [176 ms] - <7996ab5b> - Unit
[3:38:21 PM] ok 46 Chrome 71.0 - [160 ms] - <0c806320> - Unit
[3:38:21 PM] ok 47 Chrome 71.0 - [177 ms] - <c9ed76d8> - Unit
[3:38:22 PM] ok 48 Chrome 71.0 - [177 ms] - <e68ae958> - Unit
[3:38:22 PM] ok 49 Chrome 71.0 - [182 ms] - <c8ff1edb> - Unit
[3:38:22 PM] ok 50 Chrome 71.0 - [180 ms] - <1b2882a5> - Unit
[3:38:22 PM] ok 51 Chrome 71.0 - [191 ms] - <373d61f5> - Unit
[3:38:22 PM] ok 52 Chrome 71.0 - [182 ms] - <cad85f46> - Unit
[3:38:23 PM] ok 53 Chrome 71.0 - [177 ms] - <a1c4ba62> - Unit
[3:38:23 PM] ok 54 Chrome 71.0 - [180 ms] - <f976190f> - Unit
[3:38:23 PM] ok 55 Chrome 71.0 - [172 ms] - <02de6b71> - Unit
[3:38:23 PM] ok 56 Chrome 71.0 - [171 ms] - <8120379d> - Unit
[3:38:23 PM] ok 57 Chrome 71.0 - [181 ms] - <290055af> - Unit
[3:38:23 PM] ok 58 Chrome 71.0 - [182 ms] - <a14de53b> - Unit
[3:38:24 PM] ok 59 Chrome 71.0 - [174 ms] - <bcf20cdb> - Unit
[3:38:24 PM] ok 60 Chrome 71.0 - [204 ms] - <b7e497d4> - Unit
[3:38:24 PM] ok 61 Chrome 71.0 - [253 ms] - <70220201> - Unit
[3:38:24 PM] ok 62 Chrome 71.0 - [166 ms] - <9da85427> - Unit
[3:38:24 PM] ok 63 Chrome 71.0 - [165 ms] - <8381450e> - Unit
[3:38:25 PM] ok 64 Chrome 71.0 - [163 ms] - <baca70c4> - Unit
[3:38:25 PM] ok 65 Chrome 71.0 - [166 ms] - <39e63008> - Unit
[3:38:25 PM] ok 66 Chrome 71.0 - [166 ms] - <e652edf4> - Unit
[3:38:25 PM] ok 67 Chrome 71.0 - [168 ms] - <b9d552c5> - Unit
[3:38:25 PM] ok 68 Chrome 71.0 - [177 ms] - <cb8fcf7c> - Unit
[3:38:25 PM] ok 69 Chrome 71.0 - [159 ms] - <708fe3e0> - Unit
[3:38:26 PM] ok 70 Chrome 71.0 - [166 ms] - <cb900c38> - Unit
[3:38:26 PM] ok 71 Chrome 71.0 - [159 ms] - <1dcfbe49> - Unit
[3:38:26 PM] ok 72 Chrome 71.0 - [157 ms] - <ad6e6794> - Unit
[3:38:26 PM] ok 73 Chrome 71.0 - [164 ms] - <33c1b48c> - Unit
[3:38:26 PM] ok 74 Chrome 71.0 - [157 ms] - <c3605dd7> - Unit
[3:38:26 PM] ok 75 Chrome 71.0 - [149 ms] - <4d7347b2> - Unit
[3:38:26 PM] ok 76 Chrome 71.0 - [166 ms] - <a850471f> - Unit
[3:38:27 PM] ok 77 Chrome 71.0 - [182 ms] - <8481b5b4> - Unit
[3:38:27 PM] ok 78 Chrome 71.0 - [675 ms] - <26aa6a9f> - Unit
[3:38:28 PM] ok 79 Chrome 71.0 - [159 ms] - <cee276d9> - Unit
[3:38:28 PM] ok 80 Chrome 71.0 - [174 ms] - <29bf7646> - Unit
[3:38:28 PM] ok 81 Chrome 71.0 - [180 ms] - <bce2346d> - Unit
[3:38:29 PM] ok 82 Chrome 71.0 - [676 ms] - <cb78a158> - Unit
[3:38:29 PM] ok 83 Chrome 71.0 - [171 ms] - <0fa9e7a7> - Unit
[3:38:29 PM] ok 84 Chrome 71.0 - [167 ms] - <fc330e94> - Unit
[3:38:29 PM] ok 85 Chrome 71.0 - [192 ms] - <284a31da> - Unit
[3:38:29 PM] ok 86 Chrome 71.0 - [181 ms] - <88cea447> - Unit
[3:38:30 PM] ok 87 Chrome 71.0 - [808 ms] - <e43d2155> - Acceptance
[3:38:31 PM] ok 88 Chrome 71.0 - [730 ms] - <ec45e659> - Acceptance
[3:38:31 PM] ok 89 Chrome 71.0 - [239 ms] - <c85d1602> - Integration
[3:38:31 PM] ok 90 Chrome 71.0 - [242 ms] - <2fb53be6> - Integration
[3:38:32 PM] ok 91 Chrome 71.0 - [244 ms] - <db46da2a> - Integration
[3:38:32 PM] ok 92 Chrome 71.0 - [244 ms] - <99f8d090> - Integration
[3:38:34 PM] ok 93 Chrome 71.0 - [2082 ms] - <2911e01b> - Integration
[3:38:36 PM] ok 94 Chrome 71.0 - [2067 ms] - <c9e25f0d> - Integration
[3:38:38 PM] ok 95 Chrome 71.0 - [2053 ms] - <7988f7cf> - Integration
[3:38:42 PM] ok 96 Chrome 71.0 - [3867 ms] - <594b8d55> - Integration
[3:38:44 PM] ok 97 Chrome 71.0 - [2109 ms] - <020f0a6c> - Integration
```

I'm assuming the second batch of `Acceptance` are due to the work stealing algorithm? Am I understanding that correctly?

stefanpenner commented on addon-test-support/-private/ember-exam-mocha-test-loader.js:
```diff
@@ -0,0 +1,68 @@
+   * Loads the test modules depending on the urlParam
+   */
+  loadModules() {
+    debugger;
```
^^

stefanpenner commented on addon-test-support/-private/ember-exam-qunit-test-loader.js:
```diff
@@ -0,0 +1,142 @@
+      this._testem.emit('testem:get-next-module');
+    }
+
+    this._qunit.begin(() => {
```
@choheekim / @step2yeung when I remove this `this._qunit.begin(...)` the tests continue to pass, so either it is redundant or the tests are missing the scenario a where it is needed.

choheekim replied:
Working on investigating to clarify use cases of having to have `this._qunit_being(...)`

stefanpenner replied:
thanks

choheekim replied:
in ember-qunit `Ember.onerror` test is being added to QUnit module and executed. As the test is complete `moduleDone` callbacks is called and that's why we see tests continue to pass. (fyi, https://github.com/emberjs/ember-qunit/blob/faddd902d313721c33f20dd80a225946fbbd7979/addon-test-support/ember-qunit/index.js#L145) However, with `setupEmberOnerrorValidation: false` option to disable `Ember.onerror` test `moduleDone` callback is called and it results a test failing with `Browser Id 1 - global failure
    ---
        actual: >
            [object Object]
        stack: >
            Error: No tests were run.`

To prevent the incident, we have `this._qunit.begin(..)` to emit an event to emit to get a first test module to run.

stefanpenner replied:
Is there a way we can ensure there is test coverage for this behavior (or an inline comment)? The fact that it can be removed and tests pass is very risky.

choheekim replied:
I will add an inline comment to explain why `this._qunit.begin()` callback has to be registered and think of if we can write some tests against it

choheekim commented on lib/commands/exam.js:
```diff
@@ -120,141 +124,97 @@ module.exports = TestCommand.extend({
+    const ui = this.ui;
+    const events = config.custom_browser_socket_events || {};
+    const testExecutionFileName = `test-execution-${Date.now()}.json`;
+    const getBrowserId = function(testPage) {
```
The nested function is in `_addLoadBalancingBrowserSocketEvents` because we want the function to be accessible. `_addLoadBalancingBrowserSocketEvents`  is executed in testem, not in ember-exam `getBrowserId` should be within the function.

choheekim commented on addon-test-support/-private/ember-exam-qunit-test-loader.js:
```diff
@@ -0,0 +1,142 @@
+
+          // if no tests were added, request the next module
+          if (this._qunit.config.queue.length === 0) {
+            this._testem.emit('testem:get-next-module');
```
it's a test AMD module. to be more explicit what about renaming it to `testem:get-next-test-amd-module`?

scalvert commented on addon-test-support/-private/get-test-loader.js:
```diff
@@ -1,12 +1,19 @@
-  const TestLoaderModule = require(testLoaderModulePath);
-  return TestLoaderModule['default'];
-}
+  throw new Error('Cannot find known test loader.');
```
Would this error message be more accurate?

`throw new Error('Unable to find a suitable test loader. You should ensure that one of \`ember-qunit\` or \`ember-mocha\` are added as dependencies.');`

step2yeung replied:
That would. Fixed

scalvert commented on addon-test-support/-private/weight-test-modules.js:
```diff
@@ -0,0 +1,50 @@
+ * @returns {Array<string>}
+ */
+export default function weightTestModules(modules) {
+  const groups = Object.create(null);
```
Should we just be using a `Map` here?

step2yeung replied:
We can, do you see any difference between using a `Map` vs `Object.create(null);` in this case?

stefanpenner replied:
`Map` is likely more clear and concise.  As long as `Map` is available where this code will be run, I think `Map` is a good bet.

scalvert commented on addon-test-support/-private/weight-test-modules.js:
```diff
@@ -0,0 +1,50 @@
+
+  // return modules sorted by weight and alphabetically within its weighted groups
+  return Object.keys(groups)
+    .sort((a, b) =>  b-a )
```
There's some slight spacing weirdness here.

Can we do:

```js
.sort((a, b) => b - a)
```

step2yeung replied:
Fixed

scalvert commented on addon-test-support/load.js:
```diff
@@ -1,22 +1,42 @@
+ */
+function loadTests() {
+  if (testLoader === undefined) {
+    throw new Error('testLoader is not defined. It must call `loadEmberExam()` before calling `loadTest()`.')
```
Would this be more accurate?

```js
throw new Error('A testLoader instance has not been created. You must call `loadEmberExam()` before calling `loadTest()`.')
```

step2yeung replied:
Fixed

scalvert commented on lib/commands/exam.js:
```diff
@@ -120,141 +124,98 @@ module.exports = TestCommand.extend({
+    config.testPage = getMultipleTestPages(config, commandOptions);
+
+    if (commandOptions.loadBalance || commandOptions.replayExecution) {
+      config.custom_browser_socket_events = this._addLoadBalancingBrowserSocketEvents(config);
```
Any reason why this is snake case? Do we get this option from elsewhere (Testem)?

step2yeung replied:
correct, this is an option that we get from Testem.

scalvert commented on lib/utils/config-reader.js:
```diff
@@ -0,0 +1,65 @@
+ */
+function _findValidFile(files) {
+  for (let i = 0; i < files.length; i++) {
+    // TODO: investigate this cwd() usually they are in-error...
```
What does this mean?

step2yeung replied:
@choheekim do you remember what this TODO refers to again?

stefanpenner replied:
if memory serves, the `cwd`  was in the existing code. I pointed out, that it is not usually appropriate to use `cwd` in a library, rather it should be parameterized. 

But given this was existing code, merely adapted for this new case, its not the end of the world. Although I would like to encourage someone clean this up later.

choheekim replied:
I think the `TODO` has been there before this pr. I will be investigating to clean this up

scalvert commented on lib/utils/config-reader.js:
```diff
@@ -0,0 +1,65 @@
+      fs.accessSync(file, fs.F_OK);
+      return file;
+    } catch (error) {
+      debug('Failed to find ' + file + ' due to error: ' + error);
```
Can we just use a template string here?

step2yeung replied:
fixed

scalvert commented on lib/utils/config-reader.js:
```diff
@@ -0,0 +1,65 @@
+        return require(file);
+      case 'json':
+        return fs.readJsonSync(file);
+    }
```
Should this have a default case?

step2yeung replied:
fixed

scalvert commented on lib/utils/execution-state-manager.js:
```diff
@@ -0,0 +1,194 @@
+    this._replayExecutionMap = null;
+
+    // A map of browerId to test modules executed for the current test execution.
+    this._browserToModuleMap = Object.create(null);
```
Can we use a `Map` here?

step2yeung replied:
Same thing here. We could, but do you see any difference between using `Map` vs `Object.create(null)` in this case?

scalvert replied:
IMO it's readability, as `Object.create(null)` is less immediately intuitive that you're wanting `Map` functionality. With a `Map` I think it's pretty clear. Thoughts?

choheekim replied:
with a suggestion from @stefanpenner https://github.com/ember-cli/ember-exam/pull/136#discussion_r239592959 `Object.create(null)` is used here. what do you think which brings more benefits?

choheekim replied:
After off-line discussion, decided to change to use a `Map` as  a `Map` has less edge cases and  it gives more readability when neither gives no significant performance issues.

scalvert commented on lib/utils/test-page-helper.js:
```diff
@@ -0,0 +1,209 @@
+ */
+function _getFilledArray(start, end) {
+  const length = end - start + 1;
+  return Array.from({ length }, (_, i) => i + Number(start));
```
Could also use `fill` here.

step2yeung replied:
Not sure how `fill` would improve this.

scalvert commented on lib/utils/testem-events.js:
```diff
@@ -0,0 +1,165 @@
+
+      // Cleanup states from stateManager for --server mode, since a rerun of tests is triggered by
+      // a browser refresh
+      this.stateManager.cleanUpStates();
```
👍

scalvert replied:
Also, simpler.

step2yeung replied:
Makes total sense. Fixed.

scalvert commented:
This also needs a rebase.

step2yeung commented on addon-test-support/-private/ember-exam-mocha-test-loader.js:
```diff
@@ -0,0 +1,66 @@
+    this._testModules.push(moduleName);
+  }
+
+  unsee() {}
```
Here's the original override: https://github.com/ember-cli/ember-exam/blob/master/addon-test-support/-private/patch-test-loader.js#L19
From my understanding, it makes sure `this.unsee()` from [ember-cli-test-loader's Test Loader](https://github.com/ember-cli/ember-cli-test-loader/blob/master/addon-test-support/index.js#L63) does nothing. Hence it does not call `super.unsee()`.

When extending the class, does this not already override the unsee?

scalvert replied:
It will, it's just not immediately clear why we want `unsee` to noop. Does it require a comment?

step2yeung replied:
it certainly does, comment added

step2yeung commented on addon-test-support/-private/ember-exam-mocha-test-loader.js:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  /**
+   * Ember-cli-test-loader instanciates a new TestLoader instance and calls loadModules.
```
fixed

step2yeung commented on addon-test-support/-private/ember-exam-qunit-test-loader.js:
```diff
@@ -0,0 +1,142 @@
+  }
+
+  /**
+   * ember-cli-test-loader instanciates a new TestLoader instance and calls loadModules.
```
fixed

step2yeung commented on addon-test-support/-private/ember-exam-mocha-test-loader.js:
```diff
@@ -0,0 +1,66 @@
+  /**
+   * Loads the test modules depending on the urlParam
+   */
+  loadModules() {
```
Not at this moment. We have not spend time on testing what needs to be done with Mocha to get loadBalance to work.

step2yeung commented on addon-test-support/-private/get-test-loader.js:
```diff
@@ -1,12 +1,19 @@
-  if (!requirejs.entries[testLoaderModulePath]) {
-    testLoaderModulePath = 'ember-cli/test-loader';
+  if (requirejs.entries['ember-qunit/test-loader']) {
+    const TestLoaderModule = require('./ember-exam-qunit-test-loader');
```
sounds good, fixed.

stefanpenner commented on addon-test-support/-private/async-iterator.js:
```diff
@@ -0,0 +1,103 @@
+   */
+  _makeNextRequest() {
+    this._waiting = true;
+    console.debug(`makeRequest ${this._request}`);
```
we should likely not emit logging by default

stefanpenner commented on addon-test-support/-private/async-iterator.js:
```diff
@@ -0,0 +1,103 @@
+   * @param {*} response
+   */
+  handleResponse(response) {
+    console.debug(`handleResponse ${response}`);
```
we should likely not emit logging by default

stefanpenner commented on addon-test-support/-private/async-iterator.js:
```diff
@@ -0,0 +1,103 @@
+      reject = _reject;
+    });
+
+    // TODO: timeout?
```
@choheekim / @step2yeung  should we consider a timeout here? If we do not receive our response in a reasonable amount of time, it may be worth surfacing.

stefanpenner commented on addon-test-support/-private/ember-exam-qunit-test-loader.js:
```diff
@@ -0,0 +1,141 @@
+    });
+
+    const nextModuleHandler = async function(resolve, reject) {
+      try {
```
once the wrapper `new Promise` is removed, the `try/catch` and `resolve/reject` can be removed from this function

choheekim replied:
`this.loadIndividualModule` might be able to throw an error. In that case wouldn't we still want to have `try/catch`?

stefanpenner commented on addon-test-support/-private/ember-exam-qunit-test-loader.js:
```diff
@@ -0,0 +1,141 @@
+      response: 'testem:next-module-response'
+    });
+
+    const nextModuleHandler = async function(resolve, reject) {
```
Since we are using async/await, the wrapping `new Promise` is not required.

```js
 this._qunit.moduleDone(() => {
      return nextModuleHandler.call(this);
    });
```

becomes:

```js
 this._qunit.moduleDone(() => {
      return new self.Promise(nextModuleHandler.bind(this));
    });
`

stefanpenner replied:
can we also make this an arrow function, that will remove the later `bind(this)` or `call(this)

stefanpenner commented:
I've left some small nit-picks, but this is looking ready.

stefanpenner commented on addon-test-support/-private/async-iterator.js:
```diff
@@ -0,0 +1,122 @@
+      if (!this._waiting) {
+        return;
+      }
+      let err = new Error(`Promise timed out after ${this._timeout} ms while waiting for response for ${this._request}`)
```
lets prefix this error with `EmberExam: ` so its easy to track down what is is wrong.

stefanpenner commented on addon-test-support/-private/async-iterator.js:
```diff
@@ -0,0 +1,122 @@
+   * @param {*} reject
+   */
+  _setTimeout(reject) {
+    this.timer = setTimeout(() => {
```
To be safe, we should likely  clear any existing timeout, before potentially replacing it: 

```js
clearTimeout(this.timeout);
this.timeout = setTimeout(() => {
   // ...
});
```

choheekim replied:
that makes sense :)

stefanpenner commented on addon-test-support/-private/async-iterator.js:
```diff
@@ -0,0 +1,122 @@
+    this._current = {
+      resolve,
+      reject,
+      promise
```
when this promise completes, either fulfills or rejects, we must cancel the timeout.

choheekim replied:
cleanTimeout has been added to https://github.com/ember-cli/ember-exam/pull/136/files#diff-ac38a8eb1ee45cedd145f8f17a397708R57

scalvert commented on tests/test-helper.js:
```diff
@@ -13,5 +13,4 @@ Object.keys(require.entries).forEach((entry) => {
 setResolver(resolver);
-start();
-
+start();
```
This needs a newline at the end of the file.

step2yeung replied:
Fixed

scalvert commented:
Can we rebase again to pull in the upstream package changes?

step2yeung commented on lib/utils/testem-events.js:
```diff
@@ -59,12 +59,12 @@ class TestemEvents {
     const replayExecutionMap = this.stateManager.getReplayExecutionMap();
 
-    if (appMode === 'ci' && replayExecution) {
+    if (!executionOnly && replayExecution) {
```
is `executionOnly` even needed here? Because with it, setBrowserModuleQueue won't be set. Is that expected?

step2yeung commented on lib/commands/exam.js:
```diff
@@ -24,7 +24,8 @@ module.exports = TestCommand.extend({
     { name: 'replay-execution', type: String,  default: false, description: 'A JSON file path which maps from browser number to a list of modules' },
-    { name: 'replay-browser',   type: [Array, Number, String], description: 'The number of the browser(s) to run from a specified file path' }
+    { name: 'replay-browser',   type: [Array, Number, String], description: 'The number of the browser(s) to run from a specified file path' },
+    { name: 'execution-only',   type: Boolean, default: false, description: 'Disables writing a test-execution json file after running your test suite'}
```
I feel `execution-only` doesn't quite convey what the command actually do.
How about `--no-execution-file` and have an alias (short hand command) `--nef' ?

choheekim replied:
sounds right :)

step2yeung commented on lib/utils/testem-events.js:
```diff
@@ -127,13 +127,13 @@ class TestemEvents {
+  completedBrowsersHandler(browserCount, loadBalance, fileName, executionOnly) {
     this.stateManager.incrementCompletedBrowsers();
-
+    debugger;
```
`debugger` statement :)

choheekim replied:
woops..

step2yeung commented on lib/commands/exam.js:
```diff
@@ -161,12 +161,17 @@ module.exports = TestCommand.extend({
     const events = config.custom_browser_socket_events || {};
     const testExecutionFileName = `test-execution-${Date.now()}.json`;
     const getBrowserId = function(testPage) {
```
It feels like setting `command.executionOnly = true;`should be done elsewhere outside of `getBrowserId()`,  some sorta setup step. 

Do you think something like this would make sense?
```
    const getBrowserId = function() {
      // if noExecutionFile is true, the browserId should not matter, returning 0 
      if (commands.noExecutionFile) return 0;

      const browserId = /browser=\s*([0-9]*)/.exec(this.launcher.settings.test_page);

      if (Array.isArray(browserId) !== null) {
        return browserId[1];
      } else {
        throw new Error(`Browser Id couldn't be found from the test_page: ${this.launcher.settings.test_page}`);
      }
    }

...
      events['tests-start'] = function() {
        // 'tests-start' is emitted everytime a test starts.
        // To prevent from registering `processExit` and `processError` multiple times set a flag to check if callbacks are registered.
        if (!registerProcessOn) {
          if (this.process && this.launcher) {
            this.process.on('processExit', browserFailureHandler.bind(this));
            this.process.on('processError', browserFailureHandler.bind(this));
          } else {
            // if this.process or this.launcher is undefined, it means `no-launch` was passed
            // when running with `--no-launch`, a test execution file is not created
            // because Testem does not create browser instance, which is needed to correctly
            // book keep the test execution
            commands.noExecutionFile = true;
          }
          registerProcessOn = true;
        }
```

stefanpenner commented on index.js:
```diff
@@ -5,7 +5,13 @@
+  },
+
+  checkDevDependencies() {
+    const VersionChecker = require('ember-cli-version-checker');
```
@dcombs / @step2yeung  / @choheekim  this is likely causing the ember-cli-version-checker issue.

It's not clear if it is a bug in latest version-checker or a bug here.

choheekim replied:
https://github.com/ember-cli/ember-exam/pull/136/commits/8141a5daca495f4f779012e97d6b590f8df87cdc

step2yeung commented on lib/commands/exam.js:
```diff
@@ -21,14 +21,14 @@ module.exports = TestCommand.extend({
+    { name: 'parallel',             type: Number,                                   description: 'Runs your split tests on parallel child processes.' },
+    { name: 'load-balance',         type: Boolean, default: false,                  description: 'Load balance test modules. Test files will be sorted by weight from slowest (acceptance) to fastest (eslint).' },
+    { name: 'random',               type: String,  default: false,                  description: 'Randomizes your modules and tests while running your test suite.' },
+    { name: 'replay-execution',     type: String,  default: false,                  description: 'A JSON file path which maps from browser number to a list of modules' },
```
can we also have an alias here? maybe `re`

step2yeung commented on lib/commands/exam.js:
```diff
@@ -21,14 +21,14 @@ module.exports = TestCommand.extend({
+    { name: 'load-balance',         type: Boolean, default: false,                  description: 'Load balance test modules. Test files will be sorted by weight from slowest (acceptance) to fastest (eslint).' },
+    { name: 'random',               type: String,  default: false,                  description: 'Randomizes your modules and tests while running your test suite.' },
+    { name: 'replay-execution',     type: String,  default: false,                  description: 'A JSON file path which maps from browser number to a list of modules' },
+    { name: 'replay-browser',       type: [Array, Number, String],                  description: 'The number of the browser(s) to run from a specified file path' },
```
can we also have an alias here? maybe `rb`

step2yeung commented on lib/utils/tests-options-validator.js:
```diff
@@ -220,16 +216,23 @@ module.exports = class TestsOptionsValidator {
+    //When `--no-launch` option is passed, a value for launch in testem config is set to be string false.
+    if (this.options.launch === 'false') {
+      throw new SilentError(
+        'EmberExam: You must not use `no-launch` option in oorder to use the `write-execution`file` option.'
```
how about: 
'EmberExam: You must not use `no-launch` with `write-execution-file` option.'

step2yeung commented on addon-test-support/-private/ember-exam-qunit-test-loader.js:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  /**
+   * setupLoadBalanceHandlers() registers QUnit callbacks neeeded for the load-balance option.
```
typo: neeeded --> needed

step2yeung commented on lib/utils/tests-options-validator.js:
```diff
@@ -1,132 +1,339 @@
+      }
+    }
+
+    if (this.options.parallel === 0) {
```
less than 0

step2yeung commented on addon-test-support/-private/async-iterator.js:
```diff
@@ -41,7 +43,9 @@ export default class AsyncIterator {
     if (this._waiting === false) {
-      throw new Error(this.toString() + ' Was not expecting a response, but got a response');
+      throw new Error(
+        this.toString() + ' Was not expecting a response, but got a response'
```
lets use template string here
`${this.toString()} Was not expecting a response, but got a response.`

stefanpenner commented on addon-test-support/-private/ember-exam-qunit-test-loader.js:
```diff
@@ -0,0 +1,154 @@
+    });
+
+    const nextModuleHandler = () => {
+      return new Promise((resolve, reject) => {
```
nit-pick: this outer promise is redundant, and requires additional potentially error prone `resolve` / `reject` wiring.

Also, the reject(`message ${err}`) will loose `err`'s meta data. As such, it is better practice to only augment the message. That way, we don't unintentionally  lose debug information.

```js
const nextModuleHandler = () => {
  return asyncModuleIterator.next().then(response => {
    if (!response.done) {
      const moduleName = response.value;
      this.loadIndividualModule(moduleName);
      // if no tests were added, request the next module
      if (this._qunit.config.queue.length === 0) {
        return  nextModuleHandler();
      }
    }
    }).catch(e => {
      if (typeof e === 'object' && e !== null && typeof e.message === 'string') {
        e.message = `EmberExam: Failed to get next test module: ${e.message}`;
      } else {
        throw new Error(`EmberExam: Failed to get next test module: ${e}`);
      }
    });
};
``

stefanpenner commented on addon-test-support/-private/ember-exam-qunit-test-loader.js:
```diff
@@ -0,0 +1,154 @@
+    // By default ember-qunit adds `Ember.onerror` test to a qunit processing queue and once the test is complete it execute _qunit.moduleDone callback.
+    // However, when `setupEmberOnerrorValidation: false` is passed the test is disabled and _qunit.begin callback needs to request a next test module to run.
+    this._qunit.begin(() => {
+      return nextModuleHandler.call(this);
```
nit: `call(this)` is misleading as the rebound `this` will be discarded as `nextModuleHandler` is an `=>` function, which already preserves `this`


Example:
```js
a = () => console.log(this)
a(); 
> Window {postMessage: ƒ, blur: ƒ, focus: ƒ, close: ƒ, parent: Window, …} // global this

a.call('i am new this')
> Window {postMessage: ƒ, blur: ƒ, focus: ƒ, close: ƒ, parent: Window, …} // same this as before
```

stefanpenner commented on lib/commands/exam/iterate.js:
```diff
@@ -71,18 +80,21 @@ module.exports = {
   _buildForTests() {
     // TODO: execa
-    var execSync = require('child_process').execSync;
+    const execSync = require('child_process').execSync;
```
can we use `execa` here to avoid leaking child processes, as `execSync` does not cleanup child-processes always if the parent crashes or is killed.

stefanpenner commented on lib/commands/exam/iterate.js:
```diff
@@ -130,8 +142,16 @@ module.exports = {
     const chalk = require('chalk');
     const execSync = require('child_process').execSync;
```
^

choheekim commented on lib/commands/exam.js:
```diff
@@ -120,141 +129,81 @@ module.exports = TestCommand.extend({
+    const testExecutionFileName = `test-execution-${Date.now()}.json`;
+    const browserExitHandler = function () {
+      const browserCount = Object.keys(config.testPage).length;
+      testemEvents.completedBrowsersHandler(browserCount, ui, commands.get('loadBalance'), testExecutionFileName, commands.get('writeExecutionFile'));
```
Will address this in a separated pr

choheekim commented on lib/utils/execution-state-manager.js:
```diff
@@ -0,0 +1,185 @@
+   * Increment the number of completed browsers
+   */
+  incrementCompletedBrowsers() {
+    this._completedBrowsers = this._completedBrowsers + 1;
```
Will address this in a separated pr

stefanpenner commented:
this LGTM, time to merge. Release will occur once we land: https://github.com/ember-cli/ember-exam/pull/136
